sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala

```diff
@@ -1196,49 +1174,101 @@ class SQLContext(@transient val sparkContext: SparkContext)
   // Register a succesfully instantiatd context to the singleton. This should be at the end of
   // the class definition so that the singleton is updated only if there is no exception in the
   // construction of the instance.
-  SQLContext.setLastInstantiatedContext(self)
+  sparkContext.addSparkListener(new SparkListener{
+    override def onApplicationEnd(applicationEnd: SparkListenerApplicationEnd) {
```

Comment:
This is implicit `Unit`, but I think they are going to get rid of this in a future version of scala (and I think its against our style guide).